File: Algorithms_C#/20_Valid_Parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Dictionary<char, char> pairs = new Dictionary<char,char>(){
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        int n = s.Length;
        if (n % 2 == 1){
            return false;
        }
        Stack<char> stk = new Stack<char>();
        foreach (char ch in s){
            if (pairs.ContainsKey(ch)){
                if (stk.Count == 0 || stk.Pop() != pairs[ch]){
                    return false;
                }
            }
            else{
                stk.Push(ch);
            }
        }
        return stk.Count == 0 ? true : false;
    }
}

File: Algorithms_python/20_Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        pairs = {
            ")" : "(",
            "]" : "[",
            "}" : "{",
        }
        stack = []
        for ch in s:
            if ch in pairs:
                if not stack or stack[-1] != pairs[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)
        return not stack
'''
我们遍历给定的字符串 s。当我们遇到一个左括号时，我们会期望在后续的遍历中，有一个相同类型的右括号将其闭合。由于后遇到的左括号要先闭合，因此我们可以将这个左括号放入栈顶。

当我们遇到一个右括号时，我们需要将一个相同类型的左括号闭合。此时，我们可以取出栈顶的左括号并判断它们是否是相同类型的括号。如果不是相同的类型，或者栈中并没有左括号，那么字符串 sss 无效，返回 False。为了快速判断括号的类型，我们可以使用哈希表存储每一种括号。哈希表的键为右括号，值为相同类型的左括号。

在遍历结束后，如果栈中没有左括号，说明我们将字符串 sss 中的所有左括号闭合，返回 True，否则返回 False。

注意到有效字符串的长度一定为偶数，因此如果字符串的长度为奇数，我们可以直接返回 False，省去后续的遍历判断过程。
'''